### PR TITLE
[WIP] baremetal-operator: wait for Ironic services

### PR DESCRIPTION
**Pull request by andfasano**
This PR contains a PoC to address the issue related to the baremetal-operator bootstrap by the machine-api-operator: the BMO container requires that the specific Ironic services (at least an Ironic API and an Ironic Conductor) will be available before using them in the reconcile loop.

The check is performed in the BMO reconcile loop and, in case of error, the request, is rescheduled.

The waiting logic has been extracted by @stbenjam development on openshift/installer terraform provider (https://github.com/openshift-metal3/terraform-provider-ironic/pull/41) by stripping off just Terraform specific parts, and trying to keep the rest as much as possible identical.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -161,6 +161,10 @@ build: ## Build the operator binary
 tools:
 	go build -o build/_output/bin/get-hardware-details cmd/get-hardware-details/main.go
 
+.PHONY: wait-for-ironic
+wait-for-ironic:
+	go build -o build/_output/bin/wat-for-ironic cmd/wait-for-ironic/main.go
```
Comment (zaneb) [inline Makefile:166]: wat

**Comment by hardys** [comment]
I'm not sure this should be a separate command - the ironic services could become unavailable at any time (particularly in the case where the BMO is deployed with some external/existing ironic for example) - wouldn't it be cleaner if we handled this via a check at the start of the reconcile loop in the BMO itself?

File: pkg/provisioner/ironic/ironicservices.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+func waitForIronicServices(client *gophercloud.ServiceClient, inspector *gophercloud.ServiceClient, timeout int) bool {
+
+	return waitForIronic(client, timeout) &&
+		waitForInspector(inspector, timeout)
+}
+
+func waitForIronic(ironic *gophercloud.ServiceClient, timeout int) bool {
+	// Let's poll the API until it's up, or times out.
+	duration := time.Duration(timeout) * time.Second
```
Comment (dhellmann) [inline pkg/provisioner/ironic/ironicservices.go:22]: This calculation is being done everywhere. Why not make timeout a Duration type? Or move the context handling to waitForIronicServices()?

File: pkg/provisioner/ironic/ironicservices.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+// Ironic conductor can be considered up when the driver count returns non-zero.
+func waitForConductor(ctx context.Context, client *gophercloud.ServiceClient) bool {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Unable to contact API")
+			return false
+		default:
+			log.Info("Waiting for conductor API to become available...")
```
Comment (dhellmann) [inline pkg/provisioner/ironic/ironicservices.go:82]: There isn't really a conductor "API" as such, is there?

**Comment by andfasano** [comment]
Problem addressed in https://github.com/metal3-io/baremetal-operator/pull/574

**Pull request closed without merging**